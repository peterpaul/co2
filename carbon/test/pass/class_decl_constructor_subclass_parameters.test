/*
 * carbon Compiler and programming language for libco2
 *
 * Copyright (C) 2011,2012 Peterpaul Taekele Klein Haneveld
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

["stdio.h"] int printf(char * format, ...);

class ObjectA
{
	ObjectA() {
		_foo = 3;
	}

	ObjectA.foo (int _foo) {
		self._foo = _foo;
	}

	int _foo;

	int get_foo()
	{
		return _foo;
	}
}

int _round (double num) {
	return (int) (num + 0.5);
}

class ObjectB : ObjectA
{
	ObjectB(double bar) {
		super();
		self.bar = bar;
	}

	ObjectB.foo (double bar, int foo) {
		super.foo (foo);
		self.bar = bar;
	}

	double bar;

	int foobar() {
		return get_foo() + _round (bar);
	}
}

int main(int argc, char[][] argv)
{
	ObjectB a = new ObjectB(5.0);
	printf ("a.foobar() = %d\n", a.foobar());
	ObjectB b = new ObjectB.foo(5.0, 5);
	printf ("b.foobar() = %d\n", b.foobar());
	return 0;
}
