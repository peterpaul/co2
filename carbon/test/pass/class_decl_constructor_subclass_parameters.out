a.foobar() = 8
b.foobar() = 10
